File: zig/rna-transcription/rna_transcription.zig
const std = @import("std");
const mem = std.mem;

pub fn toRna(allocator: mem.Allocator, dna: []const u8) mem.Allocator.Error![]const u8 {
    var rna = try allocator.alloc(u8, dna.len);

    for (dna, 0..) |nucleotide, i| {
        switch (nucleotide) {
            'G' => rna[i] = 'C',
            'C' => rna[i] = 'G',
            'T' => rna[i] = 'A',
            'A' => rna[i] = 'U',
            else => {
                allocator.free(rna);
                return error.OutOfMemory;
            },
        }
    }

    return rna;
}

File: go/rna-transcription/rna_transcription.go
package strand

func ToRNA(dna string) string {
	var output = make([]byte, len(dna))
	for i, r := range dna {
		switch r {
		case 'G':
			output[i] = 'C'
		case 'C':
			output[i] = 'G'
		case 'T':
			output[i] = 'A'
		case 'A':
			output[i] = 'U'
		default:
			panic("Invalid DNA nucleotide")
		}
	}
	return string(output)
}
